hbase-server/src/main/java/org/apache/hadoop/hbase/rsgroup/RSGroupBasedLoadBalancer.java

```diff
@@ -203,8 +203,11 @@ public void setMasterServices(MasterServices masterServices) {
         regionList.forEach(r -> currentAssignmentMap.put(r, regions.get(r)));
         Map<ServerName, List<RegionInfo>> pairResult =
             this.internalBalancer.retainAssignment(currentAssignmentMap, pair.getSecond());
-        pairResult.forEach((server, rs) ->
-            assignments.computeIfAbsent(server, s -> Lists.newArrayList()).addAll(rs));
+        if (pairResult == null || pairResult.isEmpty()) {
```

Comment:
When will retainAssignment return null? If there is no difference between a null and an empty Map, let's just return an empty Map to avoid the null check here?

Reply:
the BaseLoadBalancer.retainAssignment may return null when online server is empty(for some situation), then will add these regions to `pendingAssignQueue` to retry assign.
actually, the RSGroupBasedLoadBalancer.retainAssignment will do `generateGroupAssignments` first, so when invoke `internalBalancer.retainAssignment` the server list can't be empty, at least has `BOGUS_SERVER_NAME`,so here can not return null

Reply:
@cuibo01 do you meet this NPE in production environment?

Reply:
Seems you do not get my point...

What I mean is what is the difference when returning null and an empty Map for the retainAssignment method... I do not care when will the method return null, I just want to know, is it OK to change to return an empty Map. If not, why?

Reply:
Yes, Let me refactor retainAssignment and roundRobinAssignment with nonNull return